src/ReactFinalForm.js

```diff
@@ -34,63 +36,128 @@ export const all: FormSubscription = formSubscriptionItems.reduce(
   },
   {}
 )
+const validationStateChanged = (a: FormState, b: FormState) =>
+  ['valid', 'invalid', 'errors'].some(key => !shallowEqual(a[key], b[key]))
 
-type State = {
-  state: FormState
-}
+const ReactFinalForm = ({
+  debug,
+  decorators,
+  destroyOnUnregister,
+  initialValues,
+  initialValuesEqual,
+  keepDirtyOnReinitialize,
+  mutators,
+  onSubmit,
+  subscription,
+  validate,
+  validateOnBlur,
+  ...rest
+}: Props) => {
+  const config: Config = {
+    debug,
+    destroyOnUnregister,
+    initialValues,
+    keepDirtyOnReinitialize,
+    mutators,
+    onSubmit,
+    validate,
+    validateOnBlur
+  }
 
-class ReactFinalForm extends React.Component<Props, State> {
-  props: Props
-  state: State
-  form: FormApi
-  mounted: boolean
-  resumeValidation: ?boolean
-  unsubscriptions: Unsubscribe[]
+  const form: FormApi = useConstant(() => {
+    const f = createForm(config)
+    f.pauseValidation()
+    return f
+  })
+  const firstRender = React.useRef(true)
 
-  constructor(props: Props) {
-    super(props)
-    const {
-      children,
-      component,
-      render,
-      subscription,
-      decorators,
-      ...rest
-    } = props
-    const config: Config = rest
-    this.mounted = false
-    try {
-      this.form = createForm(config)
-    } catch (e) {
-      // istanbul ignore next
-      if (process.env.NODE_ENV !== 'production') {
-        console.error(`Warning: ${e.message}`)
-      }
-    }
-    this.unsubscriptions = []
-    if (this.form) {
-      // set initial state
+  // synchronously register and unregister to query form state for our subscription on first render
+  const [state, setState] = React.useState<FormState>(
+    (): FormState => {
       let initialState: FormState = {}
-      this.form.subscribe((state: FormState) => {
+      form.subscribe(state => {
         initialState = state
       }, subscription || all)()
-      this.state = { state: initialState }
-    }
-    if (decorators) {
-      decorators.forEach(decorator => {
-        this.unsubscriptions.push(decorator(this.form))
-      })
+      return initialState
     }
-  }
+  )
 
-  notify = (state: FormState) => {
-    if (this.mounted) {
-      this.setState({ state })
+  const flattenedSubscription = flattenSubscription(subscription || all)
+  React.useEffect(() => {
+    // We have rendered, so all fields are no registered, so we can unpause validation
+    form.isValidationPaused() && form.resumeValidation()
+    const unsubscriptions: Unsubscribe[] = [
+      form.subscribe(s => {
+        if (firstRender.current) {
+          if (validationStateChanged(state, s)) {
```

Comment:
is it the only important thing that we can get here which should trigger rerender? 🤔

Reply:
I'm 99% sure, yes. The only thing that happens between start of form render and end of form render is that fields get registered, and the only form state that changes when that happens is potentially validation via field-level validation functions.

Timeline looks like this:

1. `Form` first render
   1. create `form`
   1. register/unregister to get initial state
   1. throw `form` into context
   1. pause validation!
   1. call render prop
1. `Field`s first render
   1. register/unregister with form to get initial state
   1. call render prop
1. `Field`'s `useEffect` fires
   1. create lasting subscription to `form`
   1. register field-level validation functions
1. `Form`'s `useEffect` fires
   1. validation unpaused and all registered field-level validation functions run
   1. create lasting subscription to `form`
   1. the only part of form state that could've changed is its validity due to the field-level validation functions

Reply:
In pure theory i.e. FormSpy can change an arbitrary thing between render & commit - so in theory something else might change

Reply:
Shit. You're right. Hence my 99%...

This was an attempt to not render twice when field level validation is used (that passes), but maybe that's just gotta be the price of using field-level validation.

Thanks for challenging me on these things. 👍 Add another 🍺 to your earnings when we meet.

Reply:
Well.... if I disable this check, the form has to render twice _always_, regardless of your validation choices...

...what if we make the rule: **if you want to modify form state on first render in a `FormSpy`, you're gonna need to put it in a `setTimeout` to trigger it later.** ?

Reply:
Gonna give it a thought some time later hopefully

Reply:
Cant we just do a shallow check here? this would cover all differences and from what I see in the source code state values which are not primitives, but objects, are having identities preserved by reusing `lastFormState.objectValue` if shallow equal pass.

Reply:
Okay, so the keys that change are the objects `modified`, `touched`, and `visited`, which, with a simple form with one `name` field, go from `{}` to `{ name: false }` on first render.

We could either:

1. special case ignore those for triggering a rerender, or
1. leave it how it is, only checking validation-related keys, or
1. just render all the damn forms twice at first

Thoughts? 🤔

Reply:
On reflection, it doesn't _have_ to be "all the damn forms", because you can select to not subscribe to those maps. But by default, it should use `shallowEqual` to compare the new and old state, and if you don't provide a subscription, your form is just gonna render twice at first.

Reply:
Seems fine to me 👌